Upgraded TinyMCE to 6.7.0

fballiano opened the pull request:
the WYSIWYG editor shipped with OM is TinyMCE 3.something, it is extremely old and outdated and has at least one known vulnerability.

example:
<img width="1111" alt="Screenshot 2023-08-07 alle 09 10 44" src="https://github.com/OpenMage/magento-lts/assets/909743/b83b650a-fbd5-4475-8b2b-a293dc2556ee">

what works (everything):
- loading/unloading the editor
- localization
- openmage-variables widget
- openmage-widgets widget
- file/image upload

note:
- there's a small modification to prototypejs in this PR, because the default one is not compatible with the new tinymce

ADDISON74 commented:
I know that some parts of OM are old and not updated, but they work. However, if it can be replaced with the latest version in next branch, I will approve it after intensive testing. 

Please note over 1k issues reported in TinyMCE repository. Here we didn't really face any reports related to this editor.

fballiano commented:
@colinmollenhour knows why this PR is here, let's just say that outdates software may bring some side effects.

ADDISON74 commented:
Let's make a visual comparison between what exists now in OpenMage and the proposed version. Even if the current one looks old, by far it has all the intuitive features for an editor. I tried to create a list just pressing a button in version 6 and I didn't succeed. It is probably hidden somewhere but does not appear in the menus.

![2](https://user-images.githubusercontent.com/8360474/234993133-4fcd6f23-02c0-4853-9472-0d32f8303d33.png)

![1](https://user-images.githubusercontent.com/8360474/234993130-d015ab3a-1975-4b44-b768-cd2c0a89af55.png)

fballiano commented:
It’s not worth testing it now, it’s extremely far from being usable

ADDISON74 commented:
The browser console must be followed. I get errors as soon as I open the editor. If I enter html tags when I press the hide/show button, the editor no longer retains the changes, errors appear in the console.

fballiano commented:
Please do 😊

empiricompany commented:
many many improvements here!
- [x] reconfigure tinymce toolbars
- [x] recreate the widget/variable plugins
- [x] recreate the image/file upload plugins

https://github.com/fballiano/openmage/pull/12

![demo](https://github.com/OpenMage/magento-lts/assets/5071467/274a530a-e0bf-46cd-a836-858cad7d08db)

In the code, there are several TODO comments and there are some choices to be made:

- the icons for the OpenMage plugins must be in SVG. For now, I have put these, I think it's fine for the variables but for the widgets, it would need to be redone, as well as the placeholder that still shows the Magento logo.
![plugins](https://github.com/OpenMage/magento-lts/assets/5071467/b39a41d6-5e65-40f1-b117-0228653cbb18)
![placeholder](https://github.com/OpenMage/magento-lts/assets/5071467/640a20d3-afe9-4e77-a1f0-5b7e81e25216)

- Localization: The localization files do not follow a standard, for example, for the French language you need to load the fr_FR.js file but for Italian you need to load the it.js file. A method could be added to detect the correct format.

https://www.tiny.cloud/get-tiny/language-packages/
https://www.tiny.cloud/docs/tinymce/6/ui-localization/

for now i have simple stripped the last 2 chars from magento locale like:

```
app/code/core/Mage/Cms/Model/Wysiwyg/Config.php:89
lang => substr(Mage::app()->getLocale()->getLocaleCode(), 0, 2)
```

then loaded in the editor:
```
js/mage/adminhtml/wysiwyg/tinymce/setup.js:89
language: this.config.lang,
```

but is not always correct

PS: need to add also in app/code/core/Mage/Adminhtml/Block/Catalog/Helper/Form/Wysiwyg/Content.php too
PSS: old editor not provide localization

- I set the dark skin for my personal choice, for me it's better because it's more focused on the content. It could also be a distinctive choice of OpenMage or it could be made configurable.
https://www.tiny.cloud/docs/tinymce/6/editor-skin/
In my opinion, we should follow the colors of this skin also for the choice of backend theme button colors, but this is a broader discussion that concerns the entire backend theme which I believe should be reviewed and made more modern without overhauling it

- The old editor loaded a CSS to format the content. I left it commented but I would remove it as it makes little sense to me.
https://www.tiny.cloud/docs/tinymce/6/add-css-options/#content_css
// content_css: this.config.content_css,
=> mage/adminhtml/wysiwyg/tinymce/themes/advanced/skins/default/content.css

There is also a CSS for popups that I think is no longer needed:
=> mage/adminhtml/wysiwyg/tinymce/themes/advanced/skins/default/dialog.css

- I tried to bring back the same old toolbar, I also added something useful plugins like emoticons and wordcount, this is also a personal choice and maybe it would be better to provide a light mode and an advanced one by adding a configuration.

- I have enabled "paste as plain text" by default, this is also a personal choice that it's better to keep enabled by default.
![paste_as_text](https://github.com/OpenMage/magento-lts/assets/5071467/5b1e4231-f17e-455e-880d-69a40b9e1d90)

- I loaded the accordion plugin which would be useful (I'm thinking of the FAQ pages) but the Javascript would need to be implemented on the frontend
![accordion](https://github.com/OpenMage/magento-lts/assets/5071467/b3dfe698-5e55-4d91-a467-5c42599944f7)

The plugin create this html markup:
```
<details class="mce-accordion" open="open">
<summary>Accordion summary...</summary>
<p>Accordion body...</p>
</details>
```

- There's probably something to clean up in the editor options, some unused variable, and in general, it remains a draft not ready for production.

fballiano commented:
@empiricompany first of all thank you for this great improvement in this PR! I wouldn't have known how to start in order to get to what you got it!

some notes:
- I'm not sure about the dark skin, I'll try it for a while but..
- I removed the content/dialog css folder
- the accordion plugin seems fine to me, it should be styled on the frontend but it's a stardard HTML element so I think it's fine as it is
- paste plain text should totally be the default
- I'll check out the localization
- I've a problem, when (inside of an editor) I click "enter" I get this error:
<img width="862" alt="Screenshot 2023-08-06 alle 09 44 02" src="https://github.com/OpenMage/magento-lts/assets/909743/7cdcffb4-4d60-4d67-810c-1c6f095de971">
have you ever seen it?

fballiano commented:
@empiricompany I've 99% solved the language issue

fballiano commented:
updated the PR description with the latest updated data

empiricompany commented:
@fballiano i can't reproduce error reported on click enter

empiricompany commented:
> updated the PR description with the latest updated data

in your first screenshot widget are not rendered as image placeholder but this actually works

empiricompany commented:
> updated the PR description with the latest updated data

> when adding a widget to the editor, a placeholder icon is shown in the editor itself. but for "variables" this is not working:

This works as before, I would keep main focus only on porting to TinyMCE 6 and postpone the new features to other pull requests for improvement

empiricompany commented:
> @fballiano I'm not sure about the dark skin, I'll try it for a while but..

i have created a new PR https://github.com/fballiano/openmage/pull/13 to allow selection of the skin, it's white by default

empiricompany commented:
> @fballiano i can't reproduce error reported on click enter

Oh, it only happens with Chrome, not with Firefox! I'll look into it

empiricompany commented:
> > @fballiano i can't reproduce error reported on click enter
> 
> Oh, it only happens with Chrome, not with Firefox! I'll look into it

seem some incompatibilities with prototype.js

![immagine](https://github.com/OpenMage/magento-lts/assets/5071467/aca1a7c8-3abc-4653-9598-7d48c72d0f2f)

tinymce want to call native function remove 
https://developer.mozilla.org/en-US/docs/Web/API/Element/remove

![tinymce](https://github.com/OpenMage/magento-lts/assets/5071467/81df6a3a-4936-4872-b225-a200f99083da)

but stuck into prototype overrided method
![protoremove](https://github.com/OpenMage/magento-lts/assets/5071467/5f44c557-2f49-47f6-8c1c-75e7ec359262)

or disable override method in object.extend resolve this issue that, for some reason, affect only chrome not firefox
![objext](https://github.com/OpenMage/magento-lts/assets/5071467/34066a82-37ed-4c51-9d63-01fbef2d9bab)

PS:
remove function is simply this that is native in modern browsers
![remove_func](https://github.com/OpenMage/magento-lts/assets/5071467/2f4c7eea-389e-4f7c-9feb-7e9bb0197b52)
Reference: https://developer.mozilla.org/en-US/docs/Web/API/Element/remove

i think we can safetly delete remove function from prototype.js? all other things seems to work correctly without this function

empiricompany commented:
> > > @fballiano i can't reproduce error reported on click enter
> > 
> > 
> > Oh, it only happens with Chrome, not with Firefox! I'll look into it
> 
> seem some incompatibilities with prototype.js
> 
> ![immagine](https://user-images.githubusercontent.com/5071467/258644663-aca1a7c8-3abc-4653-9598-7d48c72d0f2f.png)
> 
> tinymce want to call native function remove https://developer.mozilla.org/en-US/docs/Web/API/Element/remove
> 
> ![tinymce](https://user-images.githubusercontent.com/5071467/258644695-81df6a3a-4936-4872-b225-a200f99083da.png)
> 
> but stuck into prototype overrided method ![protoremove](https://user-images.githubusercontent.com/5071467/258644847-5f44c557-2f49-47f6-8c1c-75e7ec359262.png)
> 
> or disable override method in object.extend resolve this issue that, for some reason, affect only chrome not firefox ![objext](https://user-images.githubusercontent.com/5071467/258645007-34066a82-37ed-4c51-9d63-01fbef2d9bab.png)
> 
> PS: remove function is simply this that is native in modern browsers ![remove_func](https://user-images.githubusercontent.com/5071467/258645124-2f4c7eea-389e-4f7c-9feb-7e9bb0197b52.png) Reference: https://developer.mozilla.org/en-US/docs/Web/API/Element/remove
> 
> i think we can safetly delete remove function from prototype.js? all other things seems to work correctly without this function

sorry we can't delete remove function from prototype cause show this error in file browser when pick an image:
![err](https://github.com/OpenMage/magento-lts/assets/5071467/9c31c554-71fb-4e9d-bc42-25512b7e9a02)

a workaround could be wrap in a try catch, but it sucks:
![trycatch](https://github.com/OpenMage/magento-lts/assets/5071467/cfd73dc9-1877-44cf-986f-3281a57c653c)

fballiano commented:
> > when adding a widget to the editor, a placeholder icon is shown in the editor itself. but for "variables" this is not working:
> 
> This works as before, I would keep main focus only on porting to TinyMCE 6 and postpone the new features to other pull requests for improvement

@empiricompany totally correct, I didn't realize it was already that way previously

I've updated the task description with your new (merged) PR (I've switched the default skin to the tinymce-5 which really looks more consistent).

let's see how we can fix the prototypejs thing :-\

thank you so much for the work on this task!!

fballiano commented:
@empiricompany I've removed the "remove" method from prototypejs, as you suggested, and it works perfectly fine. I agree it could be removed from prototypejs since also caniuse says it's available natively on 96+% of browsers: https://caniuse.com/?search=remove

fballiano commented:
@empiricompany we could patch the "remove" method in prototypejs and make it call the native one? would it work?

empiricompany commented:
> @empiricompany we could patch the "remove" method in prototypejs and make it call the native one? would it work?

I will work on it. I want to try modifying the remove() function of the prototype by adding a check if the native function exists. For some reason, the browser file doesn't see the native function.

empiricompany commented:
@fballiano please check last commit of this PR https://github.com/fballiano/openmage/pull/14
that solve prototype error:
![removefixed](https://github.com/OpenMage/magento-lts/assets/5071467/55ff03b8-1c7f-4fed-a206-647953fbec62)

this way, both TinyMCE and file browser and other things work correctly

fballiano commented:
@empiricompany thank you for fixing prototypejs, do you think this PR is ready for general testing?

empiricompany commented:
> @empiricompany thank you for fixing prototypejs, do you think this PR is ready for general testing?

yes

ADDISON74 commented:
Great job folks. It is an important feature in OpenMage and it will have to be tested for a while with most browsers, by as many people as possible before it is merged. Please keep an eye on console

If there are no issues I think it should be merged in the main branch too.

empiricompany commented:
I have found the following issues:

- If you remove a line of the OpenMage variable, and then press the Undo button or a shortcut key, the line still contains the URL directive. For example:
Before deletion:
![a](https://github.com/OpenMage/magento-lts/assets/5071467/cf277829-18e9-462b-bb9b-79b696247ea8)
After Delete + Undo action:
![b](https://github.com/OpenMage/magento-lts/assets/5071467/84927122-3391-496f-84b0-4c81909855bb)

Error occurs also on redo action (insert openvariable + undo + redo)

Note: If you remove the OpenMage widget, it works correctly.


- The icon of the tab changed appearance on the first load. Here is an example:
![tabchanged](https://github.com/OpenMage/magento-lts/assets/5071467/9c725505-ba9a-4ecd-9603-14f9ec5839d7)

empiricompany reviewed (commented):
dismiss for newly detected issues
@fballiano please dismiss my review i can't from my side

ADDISON74 commented:
In the case of such PRs it would be good if no one were in a hurry to approve them.

I think the safest way to test is to join two windows. In one OpenMage with the current editor, in another OpenMage with this proposal. Then all the functionalities are tested step by step.

fballiano commented:
@empiricompany was that not happening with the tinymce3?

empiricompany commented:
> In the case of such PRs it would be good if no one were in a hurry to approve them.
> 
> I think the safest way to test is to join two windows. In one OpenMage with the current editor, in another OpenMage with this proposal. Then all the functionalities are tested step by step.

Sure sorry, I agree and I have always proceeded as you described. However, the recent issues reported are difficult to identify. Only by a more deep intensive testing, I was able to uncover the problem with undo/redo. As for the rest, everything seems to be functioning correctly both in the CMS area and in the EAV form used in other backend areas such as categories and products.

I recommend this tests:
- Undo/redo know issues
- Default activation/deactivation of the editor based on system configuration in General -> Content Management
- Editor functionality in CMS pages and static blocks, as well as for EAV form elements in category and product editing
- Insert / edit widget instances
- insert / edit images
- Compare the behavior between the old and new versions of the editor
(some things are bug-to-bug reported like openmage variable plugin is not rendered with a placeholder or right click in widget instance placeholder show menu option to edit image)
- Ensuring that enabling/disabling the editor maintains the correct encoding/decoding of openmage variables and widgets.
- change skins from sistem preference General -> Content Management
- change locale language must change also editor language for tootip and other gui elements (if supported by the editor)
- conduct all tests with firefox / chrome and other popular browsers / different OS


Please consider to contribute, we need help to fix last issues.

empiricompany commented:
> @empiricompany was that not happening with the tinymce3?

not in the old version of tinymce3 undo/redo on openmage directives of variables work correctly

empiricompany commented:
> > @empiricompany was that not happening with the tinymce3?
> 
> @fballiano no in the old version of tinymce3 undo/redo on openmage directives of variables work correctly

Oh yes, this also occurred in the previous version

fballiano commented:
since it was happening before it's not a new issue, it's there since 10 years and shouldn't block this PR.

empiricompany commented:
please merge this PR https://github.com/fballiano/openmage/pull/16
that fix tabs changed icon on first load

colinmollenhour commented:
Excellent work, fellas! I've done light testing and it works great!

Aside, the "insert widget" is broken in dev mode due to a PHP warning on PHP 8.2, guessing this is not related to this PR though - tested on a basically empty store so there are no widgets to insert:
![image](https://github.com/OpenMage/magento-lts/assets/38738/912cd58d-6417-4eeb-9db0-e86459c4a578)

empiricompany commented:
> Aside, the "insert widget" is broken in dev mode due to a PHP warning on PHP 8.2, guessing this is not related to this PR though - tested on a basically empty store so there are no widgets to insert: ![image](https://user-images.githubusercontent.com/38738/260600786-912cd58d-6417-4eeb-9db0-e86459c4a578.png)

This error also occurs in the current master branch on the 20.1.0 release with TinyMCE 3 for me.

I agree that we can safely merge the new editor. For me, it's perfect, but there are still a few tiny bugs that persist, just like in the older TinyMCE 3. 
Maybe we can better organize the toolbar and provide a light mode option in the configuration, but i will create a separate pull request later.

fballiano commented:
I've fixed the php8.2 bug, which wasn't really related to this PR, but should be gone now anyways ;-)

fballiano commented:
@colinmollenhour where do you think we should merge this? PR is now targeting `next`, what do you think?

fballiano commented:
I've no idea where is phpstan's error coming from (I mean, it's from the latest merge of main into next, but...)

colinmollenhour commented:
> @colinmollenhour where do you think we should merge this? PR is now targeting `next`, what do you think?

Normally I'd probably say `next`, but to a degree it is fixing a security vulnerability so if we don't merge it in main and we give the vulnerability a CVE, we will have users who cannot get the vulnerability patched without going to the next major version..

Does anyone know if there are a significant number of extensions that tap into the TinyMCE code that would be broken? If not I think this could go in main but that's just a hunch..

fballiano commented:
I've never seen a module which was modifying the wysiwyg editor but for sure somebody has it 😅

it seems right to rebase this to main, I'll try that now.

fballiano commented:
If I rebase this PR everything breaks, I could rebase with squashing but it looses the commit history and it bothers me, can we merge it in next and I backport it to main later?

also, if we take this as a security update, should we backport it to v19? it seems a bit of a stretch due to the size of this PR.

ADDISON74 commented:
If it is merged in the next branch, it has no chance of being tested in production and ready to be used when what is a next becomes main. The more we receive feedback, the more we manage to avoid any issues that arise. it is an important and necessary change, I would say let it be tested for a while and approved by at least 4 people, not necessarily green people.

fballiano commented:
I mean, I'll backport the merge commit right after... it will be the same, it will just allow me not to break the commit history of this PR.

colinmollenhour reviewed (dismissed):
Tested lightly by clicking around on the CMS page editor - clicking various buttons, inserting images, variables, saving, reloading. Verified that the security vulnerability discovered in the old TinyMCE is no longer present. Tested on PHP 8.2.

- Have not tested all places where the editor is used (product pages, etc.).
- Have not tested with complex HTML content from a production environment.
- Noticed a minor bug that must be a TinyMCE bug: jpg rotated by Windows 11 photo editor does not appear as rotated in the editor - not a blocker

Given the new editor is such a huge improvement and is probably not a critical component, there are likely to be some minor bugs found during more stringent testing but it looks good enough to me for an RC release.

fballiano commented:
I hope we can get some reviews, also, phpstan seems to fail because of its bug with php8.1, not related to this PR of the `next` branch.

elidrissidev commented:
I'll test this PR today and give my feedback.

ADDISON74 commented:
Over the weekend I will test this PR too.

ADDISON74 commented:
1 - On the first row **Font Family** dropdown is missing. We are not able to choose or change any font.

2 - On the first row **Styles** dropdown is missing too. This was used for custom styles added in a file. A very useful feature.

3 - Buttons are larger than usual. A balance must be found in relation to the other elements that already exist on the page. When the window is resized at the half on a normal viewport (1920 x 1080) a lot of vertical space is lost. You can make a visual comparison in the attached image between the one inherited from Magento and the new proposal. It is more than double. Both windows are next to each other. I don't know much about TinyMCE, but if there is an option to set the height of the icons, I would reduce it by half.

![font_family](https://github.com/OpenMage/magento-lts/assets/8360474/96b2d5e8-2ad4-4ea5-a73a-c18b76300901)

A lot of important buttons are missing

4 - **Find** and **Find and Replace** 

5 - **Insert/Edit Embedded Media** for inserting videos

6 - **Edit CSS Style**

7 - **Insert Edit Attributes**

Some of them I think are coming from plugins which are not installed.

empiricompany commented on js/mage/adminhtml/wysiwyg/tinymce/setup.js:
```diff
@@ -0,0 +1,361 @@
+
+
+    getSettings: function (mode) {
+        var plugins = 'autoresize accordion visualblocks visualchars anchor emoticons code lists advlist fullscreen pagebreak table wordcount directionality image charmap link media nonbreaking';
```
missing loading searchreplace plugin
please update 
`var plugins = 'autoresize accordion searchreplace visualblocks visualchars anchor emoticons code lists advlist fullscreen pagebreak table wordcount directionality image charmap link media nonbreaking';`

empiricompany replied:
fixed here: https://github.com/fballiano/openmage/pull/18

empiricompany commented on js/mage/adminhtml/wysiwyg/tinymce/setup.js:
```diff
@@ -0,0 +1,361 @@
+    getSettings: function (mode) {
+        var plugins = 'autoresize accordion visualblocks visualchars anchor emoticons code lists advlist fullscreen pagebreak table wordcount directionality image charmap link media nonbreaking';
+        var toolbar = [
+            'bold italic underline strikethrough | alignleft aligncenter alignright alignjustify alignnone | styles fontselect fontsize ',
```
add font family dropdown
```
 var toolbar = [
            'bold italic underline strikethrough | alignleft aligncenter alignright alignjustify alignnone | styles fontfamily fontsize',
```

empiricompany replied:
fixed here: https://github.com/fballiano/openmage/pull/18

ADDISON74 commented:
There are 6 PHPStan errors, it seems to be the same format.

fballiano commented:
> There are 6 PHPStan errors, it seems to be the same format.

phpstan is failing on next because of a phpstan bug with php8.1

ADDISON74 commented:
Another idea to improve the UI. There is too much white in the container, a solution would be to divide the buttons area from the editing area. I added a background to each toolbar and this is what came out

![toolbar](https://github.com/OpenMage/magento-lts/assets/8360474/634ffb1b-2e81-4649-9b1c-79d70f571dc9)

ADDISON74 commented:
Another UI improvement is related to the color for the first two buttons on the first row, which insert variables and widgets. Both have the wrong colors chosen, out of the schema. They must be chosen in relation to the already existing ones. In the case of the legacy theme inherited from Magento, the color must be #FFAC47, #ED6502 or #A04300. Here is how the modification looks using the second value, which I consider it the best option

![snipppets](https://github.com/OpenMage/magento-lts/assets/8360474/f29abed4-0f5d-4c2f-8af9-252c0cd65d5a)

fballiano commented:
the colors were done with the new theme in mind and they're the color of the openmage logo

pquerner commented on js/mage/adminhtml/wysiwyg/tinymce/setup.js:
```diff
@@ -0,0 +1,361 @@
+    },
+
+
+    getSettings: function (mode) {
```
Can we put each section into its own function, so its easier to overwrite if needed?

Flyingmana replied:
We can do this later in a separate PR

pquerner commented on js/mage/adminhtml/wysiwyg/tinymce/setup.js:
```diff
@@ -0,0 +1,361 @@
+ * @license     https://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+ */
+
+var tinyMceWysiwygSetup = Class.create();
```
Should we start /w prototype here or better no prototype?

empiricompany commented:
Waiting for @fballiano to merge this commit https://github.com/fballiano/openmage/pull/18 
where I fixed various things that you pointed out

> 1 - On the first row **Font Family** dropdown is missing. We are not able to choose or change any font.
> 
even though, for me, it's a bad practice, it has been added.

> 2 - On the first row **Styles** dropdown is missing too. This was used for custom styles added in a file. A very useful feature.
> 

dropdown style doesn't open anything for me in the old editor, what should it display? I couldn't find anything similar in the documentation, I think it's already incorporated in the "styles" of the new toolbar.
ref: https://www.tiny.cloud/docs/tinymce/6/available-toolbar-buttons/

> 3 - Buttons are larger than usual. A balance must be found in relation to the other elements that already exist on the page. When the window is resized at the half on a normal viewport (1920 x 1080) a lot of vertical space is lost. You can make a visual comparison in the attached image between the one inherited from Magento and the new proposal. It is more than double. Both windows are next to each other. I don't know much about TinyMCE, but if there is an option to set the height of the icons, I would reduce it by half.

What is the resolution of the screenshots? I think it's 960px? At this resolution, also other area of the backend looks messy. I don't think we can do better.
Regarding the icons, their size and colors cannot be improved unless we write a specific theme for OpenMage, which would be difficult to maintain.
there is an advanced skin and icons pack but only for premium:
https://www.tiny.cloud/docs/tinymce/6/premium-skins-and-icons/
https://www.tiny.cloud/docs/tinymce/6/small-demo/

> 
> ![font_family](https://user-images.githubusercontent.com/8360474/261815147-96b2d5e8-2ad4-4ea5-a73a-c18b76300901.jpg)
> 
> A lot of important buttons are missing
> 
> 4 - **Find** and **Find and Replace**
It was already planned in the toolbar, but the plugin had to be loaded.
in var plugins = "searchreplace"
> 
> 5 - **Insert/Edit Embedded Media** for inserting videos
added media in 
> 
> 6 - **Edit CSS Style**
> 
> 7 - **Insert Edit Attributes**
> 
i can't find related toolbar setting in the new editor for this two button (edit css style and insert edit attributes)
ref: https://www.tiny.cloud/docs/tinymce/6/available-toolbar-buttons/
But what I wonder is how much it is used and if it is useful, modern CMSs do not include all these tools in their editor.

fballiano commented:
merged PR from @empiricompany, please give it another spin

fballiano commented:
I think the font selection is useless nowadays, I'd never allow a user to select a times new roman font for a page in the store.

ADDISON74 commented:
@fballiano
> the colors were done with the new theme in mind and they're the color of the openmage logo

The use of the color proposed by you does not meet several rules, for example focusing attention on areas, which are the colors of the buttons on the page and around. If you give an image analyzer a screenshot of the backend to extract a color palette, you will see that shade of orange is among the suggested ones. The pink in the logo is not found. It must remain isolated in that logo placed at the top left of the screen. it should not be brought back to the foreground in the middle of the page because the balance is broken. I keep my opinion because I followed plastic arts, UI and UX courses.

> I think the font selection is useless nowadays, I'd never allow a user to select a times new roman font for a page in the store.

At this moment we are not in a position to decide if the font selector is no longer useful, without feedback from those who use this editor. It is very important to make the transition as smooth as possible for everyone. Personally, I don't like changes which come with issues. I would like to offer the same editor, with all the features that the user is already familiar with. If over a while some buttons want to be removed, we can have a discussion and if this is wanted by a majority they will be removed.

fballiano commented:
the original buttons had the magento colors, seems right to have the openmage colors here

ADDISON74 commented:
@empiricompany 

> What is the resolution of the screenshots? I think it's 960px? At this resolution, also other area of the backend looks messy. I don't think we can do better.

1 - Those 4 rows of toolbars must be kept fixed when reducing the width of the viewport. When two windows are attached by side, an aspect that most content editors do if they do not have two monitors, it must remain the same. As you can see in the image, we have 6 rows in the middle of 1920px.

2 - The icons must be reduced as size, I have not analyzed if it is possible adding CSS rules, but as I have added the background to the toolbars, I guess it is possible to reduce the icons as well. Basically they are SVGs. if they are like now, they pass and reach 6 rows, it takes up a lot of space vertically, in the case of certain laptops or tablets it will be a problem.

3 - There should be a custom CSS file for this editor that we can modify.

ADDISON74 commented:
> the original buttons had the magento colors, seems right to have the openmage colors here

This statement saddens me because it is not an objective argument but a subjective one.

fballiano commented:
Sorry but to me it is, i like the openmage branding and in the openmage
theme it looks nice in my opinion

Il giorno dom 20 ago 2023 alle 15:11 ADDISON ***@***.***> ha
scritto:

> the original buttons had the magento colors, seems right to have the
> openmage colors here
>
> This statement saddens me because it is not an objective argument but a
> subjective one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMage/magento-lts/pull/3220#issuecomment-1685282372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG6DL4LALGZI6AUA4PVGU3XWIEIRANCNFSM6AAAAAAXLH7LXA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ADDISON74 commented:
One more observation to analyze. If we take in consideration a popular text editor like Microsoft Word, we can see that the buttons should be arranged in a logical and efficient order. The most used closer, the less used further away.

**the first toolbar**
- Specific to OpenMage
- Insert media elements, links, anchors, specials, emoticons
- Insert HTML elements
- Actions
- Special actions like Clear formating, View source, Fullscreen

**the second toolbar**
- Undo/Redo
- Copy, Cut, Paste, ...
- Lists
- Indent
- Heading
- Sub/Superscript
- Word count
- Search

**the third toolbar**
- Bold, ...
- Alignment
- Color
- Font
- Font size
- Stiles

Below is how it looks implemented and please shrink the browser for evaluation. It is much better than it is now.

![toolbars](https://github.com/OpenMage/magento-lts/assets/8360474/db684906-6d75-49e4-946c-db627621aff5)

This is the code /js/mage/adminhtml/wysiwyg/tinymce/setup.js

```js
var toolbar = [
            'table image media link unlink anchor charmap emoticons | blockquote hr nonbreaking | visualblocks visualchars pagebreak | ltr rtl |  removeformat code fullscreen ',
            'undo redo | copy cut paste pastetext | bullist numlist advlist accordion | indent outdent | h1 h2 h3 h4 h5 h6 |  subscript superscript | wordcount | searchreplace',
            'bold italic underline strikethrough | alignleft aligncenter alignright alignjustify alignnone | forecolor backcolor | fontfamily | fontsize | styles'
        ];
```

ADDISON74 commented:
Also, if there is no immediate solution to reduce the size of the buttons, the removal of the group dividers should be taken into account. When the size of the browser window is reduced, moving the buttons to new rows will no longer be so visually unpleasant.

Here's how it looks without dividers. My opinion is that it looks much better than the current version.

![toolbars_delimiters](https://github.com/OpenMage/magento-lts/assets/8360474/db643e54-9a55-433d-b01a-1dc7ee8b1913)

fballiano commented:
i've removed the colors from the icons and i've removed the emoticons plugin, since we don't have a mb4 database the emoticons won't be saved.

ADDISON74 commented:
So far in use I have not encountered any problems and I would like to thank those who contributed to have such a result.

Changes are clearly needed related to the way the editor is displayed in relation to the user (UI and UX), the features offered (for example options such as auto-resize, a few optimal plugins for better speed, themes for legacy and openmage). The user manual is very well done and I think I can propose a lot of useful changes.

For a start I would leave with a menu, a single toolbar and no dots. Thus eliminating a lot of problems that have been reported (including mine). The implementation is also given as an example in the manual. This is what the editor would look like for the beginning, then later we can add features for OpenMage. We can evaluate a context menu. Menus with icons are looking great. If we decide to use more than one toolbar we can collapse the others and use the dots feature.
*************
**INITIAL LOOK** - it take half of the current vertical space which is a good improvement.

![tiny1](https://github.com/OpenMage/magento-lts/assets/8360474/6daacb2c-3a71-4442-8e4d-30c321bcfb8c)
*************
**USING DOTS** - if more than one toolbar is needed.

![tiny2](https://github.com/OpenMage/magento-lts/assets/8360474/0e328bc0-f7e0-4e3b-93a5-62a265488124)
*************
**Menus**

![menus](https://github.com/OpenMage/magento-lts/assets/8360474/25845153-7f70-4a38-ab9f-4498fc052a62)

fballiano commented:
@ADDISON74 I don't understand which layout you'd like, one single toolbar? it was the original implementation but you asked us to have the 3 toolbars as it is now coded in this PR.

empiricompany commented:
@ADDISON74 - Honestly, I had tried the solution with the dots but I hadn't succeeded.

ADDISON74 commented:
@fballiano - Please take a look over my post from the end of April to inform better what I reported at that time. There was a great lack of features (see the menus, the toolbar). You replied me it is a draft and I didn't comment anymore.

Even if I had asked for an implementation that was taken into account, if I was wrong, I have the right to change my opinion. Up to this moment I can say without any hesitation that no opinion of mine related to this PR has been taken into account. The fact that my posts are ignored, that I am offered subjective arguments, made it clear to me to stay in my place. I can live with that, I will no longer bother with my opinions on this PR. 

Anyway I consider that this change must benefit from the best in order to be accepted by the majority. What I posted these days comes after testing in the last week. If I knew that I would be treated like this, I would allocate my time in other directions. And most importantly, I wanted to bring it to the main branch as well.

I hope you will appreciate my honest opinion.

ADDISON74 commented:
@empiricompany  - I will analyze the use of dots.  The settings in the JS file are here https://www.tiny.cloud/docs/tinymce/6/full-featured-open-source-demo/.

empiricompany commented:
> @empiricompany - I will analyze the use of dots. The settings in the JS file are here https://www.tiny.cloud/docs/tinymce/6/full-featured-open-source-demo/.

yes i know but i can't figure it out

ADDISON74 commented:
@empiricompany - Please visit the link and select **Edit on CodePen**. I have no idea what options the author set, but I was able to quickly notice two things

- all the buttons are in one toolbar only
- when the window is narrowed as width, a part of the buttons will be displayed and the rest are replaced with dots

fballiano commented:
@ADDISON74 I can assure you none of your notes were dismissed, personally I had a couple of weeks of holiday (first time in 10 years) and before I was working on many PRs and, on this one, I was totally stuck, I couldn't make it work so that's why it was on hold for a while. If it wasn't for @empiricompany that magically fixed everything...

But I wan't to point out some things:
- nobody here must do what other people say, everybody respect everybody's opinion but that doesn't mean that they are orders, at the end of the day OM it's a fun project for everybody and nobody is anybody's boss, we're all the same, that's why we vote for things :-)
- everybody can change opinions, sure, but sometimes we may judge too quickly, for example in this PR we spent many, many, many hours and it's not nice to read "this is wrong" kind of comments you know, when later in a few months you also recognize that your opinion changed

there's really no drama, we all want to make OM better

empiricompany commented:
> @empiricompany - Please visit the link and select **Edit on CodePen**. I have no idea what options the author set, but I was able to quickly notice two things
> 
>     * all the buttons are in one toolbar only
> 
>     * when the window is narrowed as width, a part of the buttons will be displayed and the rest are replaced with dots

@ADDISON74 - i think this is the correct documentation: https://www.tiny.cloud/docs/tinymce/6/toolbar-configuration-options/#toolbar_mode
i tried to set toobar_mode sliding but nothing change

ADDISON74 commented:
@empiricompany - I will investigate which are the most suitable/optimal options for the menu and tool, in which order, including the dots button. Even if they won't be accepted in OpenMage, at least I'll put them somewhere public for whoever wants to use them.

fballiano commented:
toolbar_mode "scrolling" works also with 3 toolbars, it's not great... but probably better than having 6 rows of buttons...

<img width="773" alt="Screenshot 2023-08-21 alle 12 56 40" src="https://github.com/OpenMage/magento-lts/assets/909743/95bb9849-12c8-4dcc-a64a-6d4a4076f2c2">

empiricompany commented:
I don't have a lot of time to dedicate to this task and perform various tests on the toolbar configuration, but I believe what we have now is stable and ready for review to identify any important blocking issues, and then merge it soon

@ADDISON74, it would be better if you contributed through pull requests instead of just describing what you want to achieve. I would be happy to test them.

In my opinion, we should make an initial release as it is, exactly how it was before. **It would be already a big change**. Then, everyone can open additional PR to improvement it. Also, in my opinion, the size and colors of the icons are fine as they are, as well as the little issues at low resolutions and additional tool like "Edit CSS Style" and "Insert Edit Attributes" reported by @ADDISON74, are not that important to me.

It would be nice if other partecipants also shared their opinion on this pull request.

empiricompany commented:
> toolbar_mode "scrolling" works also with 3 toolbars, it's not great... but probably better than having 6 rows of buttons...
I tested it, it's not the best but it can be a good compromise for now

fballiano commented:
I've enabled the scrolling toolbar, which is the only one working at the moment. not the best for sure, but 6 rows are too awful, that's true.

if we then redesign the toolbars then we'll see.

I probably would have preferred this kind of approach:
<img width="998" alt="Screenshot 2023-08-21 alle 13 27 21" src="https://github.com/OpenMage/magento-lts/assets/909743/ce333f5d-d55f-423c-9306-a72dc3e9d4dd">

but I agree with @empiricompany that if it's working then we should try to merge it soon so that we'll be able to get more feedback from the rest of the community :-)

fballiano commented:
> > toolbar_mode "scrolling" works also with 3 toolbars, it's not great... but probably better than having 6 rows of buttons...
> 
> I tested it, it's not the best but it can be a good compromise for now ![toollbar_mode_scrolling](https://user-images.githubusercontent.com/5071467/262022551-532719c5-05bb-4759-957b-71de81a9adcd.png)

wow, I don't see it that ugly on mac 😅😅😅😅

empiricompany commented:
> I've enabled the scrolling toolbar, which is the only one working at the moment. not the best for sure, but 6 rows are too awful, that's true.
> 
> if we then redesign the toolbars then we'll see.
> 
> I probably would have preferred this kind of approach: <img alt="Screenshot 2023-08-21 alle 13 27 21" width="998" src="https://user-images.githubusercontent.com/909743/262022860-ce333f5d-d55f-423c-9306-a72dc3e9d4dd.png">
> 
> but I agree with @empiricompany that if it's working then we should try to merge it soon so that we'll be able to get more feedback from the rest of the community :-)

I like this toolbar configuration, it could be proposed later in another improvement PR

fballiano commented:
it's the default here: https://www.tiny.cloud/tinymce/features/ I think it was my first implementation (more or less)

ADDISON74 commented:
Please take your time. As it was until a few months ago, it can wait for a while You worked with two people, now you also have me here, even if only ideas come. I want the be merged in a shape attractive and optimal for editing. In addition, I want it in the main branch. from that point forward, improvements can be applied through PRs, but I don't want to end up working in two branches again. I'm not German, but they have a saying, "measure 10 times, cut only once".

So, the backend seems to be fine, it just needs to be checked. The frontend side has issues and we can solve them. It's like you have the data, you put it in HTML format, but you don't apply any CSS. It is natural to look unattractive. TinyMCE is powerful and it has many features to consider.

This post is not an order, but only a personal opinion.

ADDISON74 commented:
It is fine, I am not opposed to this PR staying in the next branch for a while until it is suitable for main too. I identified what the problems are and I think they can be solved.

ADDISON74 commented:
https://www.tiny.cloud/docs/tinymce/6/full-featured-open-source-demo/

The cool result is that in this scenario (1 menu + 1 toolbar), reducing the browser window even to the minimum will keep the same vertical space. The dots menu will take over the rest of the buttons regardless of how small the window is (this is not the case because the OpenMange Backend is not responsive). Also, it should be noted that it has delimiters between the groups of buttons, but they are not visible - a white ball for this implementation and an elegant implementation without abusing of borders.

fballiano commented:
I've just committed a new version that uses this layout:

<img width="782" alt="Screenshot 2023-08-22 alle 15 06 13" src="https://github.com/OpenMage/magento-lts/assets/909743/5d68a7b1-0c2b-4abb-b203-0f4f7a2105e5">

Note: Adding the "openmage variable" and "openmage widget" menu item (under the "insert" menu) it's not that clean because I had to override the menu configuration but it seems to me that there's no other way of doing it

ADDISON74 commented:
@fballiano - It looks much better. I will provide a feedback.

fballiano commented:
Fixed the label and the widget insert from the menubar (which doesn't support the active state, that's why there was that js error).

I don't see the blurry icon but I've always notices that one of the SVG you added had a size of 18px and the other of 24, dunno if it comes from that.

Because we're a small group (of the active people) it often happens that one says X, the other says Y and then PRs get stuck forever... so I don't know what to do but these stalling situations bother me a lot.

empiricompany commented:
> Fixed the label and the widget insert from the menubar (which doesn't support the active state, that's why there was that js error).
> 
> I don't see the blurry icon (neither in chrome nor firefox) but I've always notices that one of the SVG you added had a size of 18px and the other of 24, dunno if it comes from that.
> 
> Because we're a small group (of the active people) it often happens that one says X, the other says Y and then PRs get stuck forever... so I don't know what to do but these stalling situations bother me a lot.

ok for me can be merged no bother
this commit fix openmagevariable icon 
https://github.com/fballiano/openmage/pull/19

fballiano commented:
merged your PR, we're back on track :-)

ADDISON74 commented:
My opinion is to give it a shape starting from some initial conditions. The editor should have a menu, a responsive toolbar, optimal plugins.

We check the initial implementation then we approve it. After that, in individual PRs we take care of the menu, toolbar, extra options, themes for each Backend (Magento Legacy and OpenMage). For each visual part, I would like to discuss first and then move on to implementation, it is more efficient for everyone that way, because there will be consensus.

If we find something interesting through the documentation to be added, we will discuss. It doesn't have to be the Christmas tree, it doesn't have to be heavy loaded, but not the ugliness of the earth. I like this editor and even if I haven't used it much in the past, many will find it useful.

pquerner commented:
A goal for me, if we are to give an upgraded RTE, give sane defaults but make it that its easily customizable. Thats what I always liked from TYPO3.

However, one has to argue that it might be not too high priority, since "who wants it" anyways.
(Or: its open source now, fix it and add a PR ;P)

colinmollenhour commented:
Please keep in mind this isn't a forum for feature requests, this is Fabrizio's PR. He set out to simply drop some extremely dated code which is no longer supported and replace it with a newer version that is supported by a vendor. The original motivation is that there was found a security vulnerability in TinyMCE 3 and the only/best/feasible way to fix is to upgrade to TinyMCE 6.

Since Fabrizio is investing his time on this task I believe he should be given great latitude on this PR. He has been very receptive of constructive criticism and you all have good feedback, but I think we need to be respectful of his time and try very hard not to let feature creep and bike shedding set in. Let's help him make sure there are no major regressions in functionality and no security holes and get it merged and then I think a separate issue/discussion should be opened to discuss UX issues. For example, open a GH discussion with some proposals and mockups and questions for feedback. This could also be done in parallel, just not on this PR in my opinion.

ADDISON74 commented:
Go to **Help** menu and select the only link. There is no window displayed as it should be. In the browser console appears the following error. 

![console](https://github.com/OpenMage/magento-lts/assets/8360474/929ec416-8252-43c1-9ca4-b0ad0c8261ab)

empiricompany commented:
different error on Firefox
![help_plugin](https://github.com/OpenMage/magento-lts/assets/5071467/b4551787-f08a-4623-9b61-8357d87b9e44)

fballiano commented:
> @fballiano - fixed here [fballiano#20](https://github.com/fballiano/openmage/pull/20)

amazing!! merged already

ADDISON74 commented:
I confirm the bug related to the Help menu was fixed. 

I will spend a week or two with this PR, after which I propose to integrate it into the next branch. Later we can discuss about menus, buttons, toolbar, status bar, themes and many others that can be solved later with PRs. When nothing can be added or removed, that is, it is perfect, it can be taken to the main branch. I think we can achieve this by the end of the year.

Not to forget an important aspect. how will we do periodic updates?

fballiano commented:
I think we either merge it to both next and main now or it will be impossible to backport multiple PRs from next to main in the future.

fballiano commented:
and I remember we all agreed to merge this to main the moment it gets approved, I wouldn't change it now, again.

fballiano commented:
updates will have to be managed manually at the moment.

ADDISON74 commented:
If we change the branch to main, isn't it better? All future PRs will be in main and will be in next too.

fballiano commented:
> If we change the branch to main, isn't it better? All future PRs will be in main and will be in next too.

I wrote it a few comments ago, it I move the PR to the main branch it breaks, please, let's leave it like this.
I constantly merge all main->next every few days so it's not gonna be a problem.

fballiano commented:
What do we have to get people to try this PR?
It fixes a vulnerability (this aspect it not weighted enough) and it updates one of the oldest component of the backend and makes it look good again with something like 10 years of improvements from upstream.

ADDISON74 commented:
At the end of the week I approve it, then I will propose discussions in the Ideas section to agree on some necessary changes, after which we implement them. Especially on the UX/UI side, so that everyone welcomes the change.

pquerner commented:
Perhaps we can setup a small [Cypress](https://www.cypress.io/) test suite which checks the RTE?
The output HTML shouldnt change much, right? (Or should it?)

pquerner commented:
I know Cypress, I've used it previously. Its basically a way to control a browser from javascript.
I'll setup some small repo and let you guys know.

In my mind, a test case could look like this:

Test the Old RTE 
- Paste some text
- Mark it red
- Did it become red?
- Test green

Do the same with the new RTE. 

And then come up with more test cases (especially the Magento-Only features, such as Widgets). Checking the HTML output is crucial imho. People expect their CMS pages to look the same even after clicking save on the CMS page without touching anything else.

fballiano commented:
and I have to emphathize again that tinymce3 has an open vulnerability, we have to upgrade one way or the other

pquerner commented:
> I guess the first part of the tests would actually be for the tinymce repository, I'm absolutely sure that tinymce works as it should already :-) for sure not worse than 10 years ago :-D

Okay, so I should only write tests for magento-only features?
I wouldnt test ALL the basics of this RTE, of course. That shouldnt be needed.

empiricompany commented:
for me, it's perfect. I'm already using it with several clients, and they're really satisfied about it

colinmollenhour commented:
A test suite would be great (and Cypress is an excellent choice) but to me that's a new feature so I think it should not become a blocker for this task. My only stipulations would be that if there is a test suite it must run automated via Github Actions and be easy to setup and run locally.

@pquerner I propose you create a new branch based on the one in this PR and add the Cypress test suite there and if it is done before this branch is ready to merge, great, they can both be merged together. However, if it is not ready in time, no problem, it can be merged later.

It appears TinyMCE doesn't really have their own test suite, but still I think just making sure the basic functionality (turn WYSIWYG off/on) and the OM add-ons (images, variables, widgets) work as expected would be sufficient. If there are changes between TinyMCE 3 and 6 for normal content they will probably not become apparent until it gets broader use and not every difference will indicate there is a regression as we may decide the new behavior is preferred anyway.

So is PHPStan the only blocker now for merging?

fballiano commented:
@colinmollenhour If I'm not mistaken it's a PHPStan bug with PHP 8.1, nothing to so with this PR, the whole `next` branch shows this error (regardless of this PR which is not merged yet)

colinmollenhour commented:
Attempting to fix PHPStan with #3498

fballiano commented:
all green checks again ;-)

fballiano commented:
It is a bit frustrating not to get a single review for such a big effort mostly @empiricompany put into this.

we prefer keeping a >10 years old component with a published vulnerability instead of this.

empiricompany commented:
> It is a bit frustrating not to get a single review for such a big effort mostly @empiricompany put into this.
> 
> we prefer keeping a >10 years old component with a published vulnerability instead of this.

yes we should merge it ASAP this PR is perfect, i have dedicated a lot of time to it and also fixed all minor things ..
IMHO in general, we should speed up the development of OpenMage and move forward quickly; otherwise, we will miss the opportunity to become a independent and affidable e-commerce platform

colinmollenhour reviewed (approved):
Awesome work, guys! I've tested in Chrome just clicking around, adding variables, widgets and images and saving an reloading and switching WYSIWYG off and on.

I can't find any issues with it, works beautifully!

pquerner commented:
I've started the cypress tests here btw: https://github.com/pquerner/openmage_cypress_tests 
In case anyone wants to take a look.

I think I write a few tests for widget and magento-only functionality and that should be it.

fballiano commented:
@kiatng @elidrissidev @ADDISON74 @Flyingmana @tmotyl sorry to bother you but we would need to get some review on this PR.

notes:
- there's an public vulnerability on the version of tinymce that's currently in OM
- this PR is for `main` and I'll backport it as soon as it gets merged

Flyingmana commented:
I will need maybe 2 days to fully review the changes.


An advise. If something is of importance for security, keep the amount of changed lines and files low 
No codestyle changes.
If the framework/library changes its used variable name, link the previous one to it so you dont need to rename all.

ADDISON74 commented:
I've been testing it in production for about 2 weeks and I haven't identified any problems. I haven't approved it yet because I have about 5% left to be fully convinced. I also received quite useful feedback. There is room for other changes, but they can be made in time and in separate PRs. In general, these changes are TinyMCE features for which the API must be read in full.

fballiano commented:
about workflow failing after https://github.com/OpenMage/magento-lts/pull/3536, I think https://github.com/OpenMage/magento-lts/pull/3542 would fix. anyway they don't fail because of this PR, it's the `next` branch failing, not the PR.

Flyingmana commented:
These files still use the `tinyMCE` writing

* app/design/adminhtml/default/default/template/widget/form/element.phtml
* js/mage/adminhtml/browser.js
* js/mage/adminhtml/wysiwyg/widget.js

Flyingmana commented:
None of the things I commented I would count as blocking because of the security relevance, but maybe we should look again at them and if needed fix them later then

fballiano commented on js/mage/adminhtml/variables.js:
```diff
@@ -112,7 +112,7 @@ var Variables = {
 };
 
-MagentovariablePlugin = {
+OpenmagevariablePlugin = {
```
actually the naming changes were done to make it easier to find something like what you highlight in https://github.com/OpenMage/magento-lts/pull/3220#issuecomment-1741856431 actually I'll check them out right now, thank you for noticing!!!

fballiano commented on js/mage/adminhtml/browser.js:
```diff
@@ -226,19 +223,13 @@ Mediabrowser.prototype = {
-         } else {
-             return null;
-         }
+    getMediaBrowserCallback: function() {
```
well initially my idea was to have it in the `next` branch because, since all tinymce's variables/objects have different names between v3 and v6... everything is BC so having better names for our method didn't make any change.

later it was sad to have it into `main` which changes things a bit but at the end of the day, all of tinymce's scripts are still incompatible so... also if we don't change our naming... we don't get much compatibility anyway.

what do you think?

fballiano commented on app/code/core/Mage/Widget/controllers/Adminhtml/WidgetController.php:
```diff
@@ -34,10 +34,11 @@ public function indexAction()
         // save extra params for widgets insertion form
         $skipped = $this->getRequest()->getParam('skip_widgets');
-        $skipped = Mage::getSingleton('widget/widget_config')->decodeWidgetsFromQuery($skipped);
+        if (is_string($skipped)) {
```
in order not to drag this PR any longer I'd address it in a separate PR ASAP :-)

fballiano commented:
is it ok if I merge and port to `main` now? then work on @Flyingmana's notes in a PR on the `main` branch?

ADDISON74 commented:
I looked over this PR this weekend too. There are still changes to be made, but they are not a priority because all are related to with UI/UX. The editor is functional and ready for production.

Could we extract the version we are replacing in a repository? If someone is not happy with this new version and wants the previous one, he can replace it on his own responsibility.

I agree with porting to the main branch and we continue the work from there. All future PRs will go to next too.

fballiano commented:
@ADDISON74 create something like an extension to return to tinymce v3 is complex because it's like 500 files, with phtml in many different places, I'm sorry but I really wouldn't do that, it's too much work and I can't dedicate that time. anybody could do it if somebody feels the need for that, although highly un-suggested because of the security implications.

fballiano commented:
oki guys it's time, I'll merge and port to `main` so we can continue there, I feel that after ZF1F this is another major stepstone!

colinmollenhour commented:
Nice work, @fballiano! Also thanks for everyone else's testing and feedback!
